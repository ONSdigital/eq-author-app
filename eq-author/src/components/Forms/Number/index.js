import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { clamp, isNaN } from "lodash";

import Input from "components/Forms/Input";
import { NUMBER, PERCENTAGE, CURRENCY, UNIT } from "constants/answer-types";
import { unitConversion } from "constants/unit-types";
import { radius } from "constants/theme";

const StyledDiv = styled.div`
  display: flex;
  flex-direction: row;
  position: relative;
  align-items: center;
  width: 8em;
`;

export const NumberInput = styled(Input)`
  width: 100%;
  border-radius: ${radius};
  outline: none;

  &[type="number"]::-webkit-inner-spin-button,
  &[type="number"]::-webkit-outer-spin-button {
    appearance: none;
    margin: 0;
  }

  &[type="number"] {
    appearance: textfield;
  }

  ${(props) => props.valueType === CURRENCY && "padding-left: 1.4em"}
  ${(props) => props.valueType === PERCENTAGE && "padding-right: 1.6em"}
`;

const UnitSymbol = styled.div`
  position: absolute;
  opacity: 0.5;
  ${(props) => (props.trailing ? "right: 0.5em" : "left: 0.5em")}
`;

const Number = (props) => {
  const {
    id,
    onChange,
    onBlur,
    value,
    className,
    type,
    unit,
    name,
    min,
    max,
    step,
    ...otherProps
  } = props;
  const unitId = `unit-${id}`;
  const handleChange = ({ value }) => {
    if (value.length === 0) {
      onChange({ name, value: null });
      return;
    }

    const enteredValue = clamp(parseInt(value, 10), min, max);
    const newValue =
      isNaN(enteredValue) || Object.is(enteredValue, -0)
        ? props.default
        : enteredValue;
    onChange({ name: name || id, value: newValue });
  };

  const handleBlur = () => {
    if (value === null) {
      onChange({ name, value: props.default });
    }
    if (onBlur) {
      setTimeout(onBlur);
    }
  };

  return (
    <StyledDiv className={className}>
      <NumberInput
        id={id}
        data-test={props["data-test"]}
        value={value}
        onChange={handleChange}
        type="number"
        onBlur={handleBlur}
        aria-live="assertive"
        valueType={type}
        aria-labelledby={unitId}
        min={min}
        max={max}
        default={props.default}
        name={name}
        step={step}
        {...otherProps}
      />
      {type === CURRENCY && (
        <UnitSymbol id={unitId} data-test="unit">
          Â£
        </UnitSymbol>
      )}
      {type === PERCENTAGE && (
        <UnitSymbol id={unitId} data-test="unit" trailing>
          %
        </UnitSymbol>
      )}
      {type === UNIT && (
        <UnitSymbol id={unitId} data-test="unit" trailing>
          {unit && unitConversion[unit].abbreviation}
        </UnitSymbol>
      )}
    </StyledDiv>
  );
};

Number.defaultProps = {
  min: 0,
  step: 1,
  default: null,
  "data-test": "number-input",
};

Number.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onBlur: PropTypes.func,
  value: PropTypes.number,
  step: PropTypes.number,
  min: PropTypes.number,
  max: PropTypes.number.isRequired,
  default: PropTypes.number,
  type: PropTypes.oneOf([CURRENCY, PERCENTAGE, NUMBER, UNIT]),
  "data-test": PropTypes.string,
  className: PropTypes.string,
  unit: PropTypes.string,
};

export default Number;
