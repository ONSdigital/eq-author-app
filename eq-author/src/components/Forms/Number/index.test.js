import React from "react";
import { shallow } from "enzyme";
import Number from "./";

const defaultValue = 0;

const afterImmediate = () => new Promise(resolve => setImmediate(resolve));

describe("Number", () => {
  let number;
  let numberWithMinMax;
  let handleChange;
  let handleBlur;

  beforeEach(() => {
    handleChange = jest.fn();
    handleBlur = jest.fn();
    number = shallow(
      <Number
        name="numberName"
        id="number"
        onChange={handleChange}
        onBlur={handleBlur}
        value={defaultValue}
        min={-10}
        max={10}
      />
    );
    numberWithMinMax = shallow(
      <Number
        id="number"
        onChange={handleChange}
        value={defaultValue}
        min={0}
        max={100}
      />
    );
  });

  it("should render correctly", () => {
    expect(number).toMatchSnapshot();
  });

  it("should update UI when changed", () => {
    number.find("[data-test='number-input']").simulate("change", { value: 3 });
    expect(number.find("[data-test='number-input']").prop("value")).toEqual(3);
  });

  it("should call onChange when blurred", () => {
    number.find("[data-test='number-input']").simulate("change", { value: 3 });
    number.find("[data-test='number-input']").simulate("blur");
    expect(handleChange).toHaveBeenCalledWith({
      name: "numberName",
      value: 3
    });
  });

  it("should trigger change and then blur when blurred", cb => {
    number.find("[data-test='number-input']").simulate("change", { value: 5 });
    number.find("[data-test='number-input']").simulate("blur");
    expect(handleChange).toBeCalled();
    return afterImmediate()
      .then(() => {
        expect(handleBlur).toBeCalled();
      })
      .then(cb)
      .catch(cb);
  });

  it("should not call blur if blur not provided", cb => {
    numberWithMinMax
      .find("[data-test='number-input']")
      .simulate("change", { value: 5 });
    numberWithMinMax.find("[data-test='number-input']").simulate("blur");
    expect(handleChange).toBeCalled();
    return afterImmediate()
      .then(() => {
        expect(handleBlur).not.toBeCalled();
      })
      .then(cb)
      .catch(cb);
  });

  describe("min/max", () => {
    it("should not go over the max when changed", () => {
      numberWithMinMax
        .find("[data-test='number-input']")
        .simulate("change", { value: 101 });
      numberWithMinMax.find("[data-test='number-input']").simulate("blur");
      expect(handleChange).toBeCalledWith({
        name: "number",
        value: 100
      });
      expect(
        numberWithMinMax.find("[data-test='number-input']").prop("value")
      ).toEqual(100);
    });

    it("should not go under the min when changed", () => {
      numberWithMinMax
        .find("[data-test='number-input']")
        .simulate("change", { value: -1 });
      numberWithMinMax.find("[data-test='number-input']").simulate("blur");
      expect(handleChange).toBeCalledWith({
        name: "number",
        value: 0
      });
      expect(
        numberWithMinMax.find("[data-test='number-input']").prop("value")
      ).toEqual(0);
    });

    it("should default to min when NaN", () => {
      numberWithMinMax
        .find("[data-test='number-input']")
        .simulate("change", { value: " " });
      numberWithMinMax.find("[data-test='number-input']").simulate("blur");
      expect(handleChange).toBeCalledWith({
        name: "number",
        value: 0
      });
      expect(
        numberWithMinMax.find("[data-test='number-input']").prop("value")
      ).toEqual(0);
    });
  });
});
