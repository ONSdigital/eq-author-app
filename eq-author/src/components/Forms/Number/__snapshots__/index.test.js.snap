// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number min/max render with a unit type for currency 1`] = `
<Number__UnitSymbol
  data-test="unit"
  id="unit-number"
>
  Â£
</Number__UnitSymbol>
`;

exports[`Number min/max render with a unit type for percentage 1`] = `
<Number__UnitSymbol
  data-test="unit"
  id="unit-number"
  trailing={true}
>
  %
</Number__UnitSymbol>
`;

exports[`Number min/max render with a unit type for unit 1`] = `
<Number__UnitSymbol
  data-test="unit"
  id="unit-number"
  trailing={true}
>
  cm
</Number__UnitSymbol>
`;

exports[`Number should render correctly 1`] = `
<Number__StyledDiv>
  <Number__NumberInput
    aria-labelledby="number"
    aria-live="assertive"
    data-test="number-input"
    default={null}
    id="number"
    max={10}
    min={-10}
    name="numberName"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    step={1}
    type="number"
    value={null}
  />
</Number__StyledDiv>
`;
